ipn/store/kubestore/store_kube.go

```diff
@@ -84,27 +116,79 @@ func (s *Store) ReadState(id ipn.StateKey) ([]byte, error) {
 
 // WriteState implements the StateStore interface.
 func (s *Store) WriteState(id ipn.StateKey, bs []byte) (err error) {
-	return s.updateStateSecret(map[string][]byte{string(id): bs})
+	defer func() {
+		if err == nil {
+			s.memory.WriteState(ipn.StateKey(sanitizeKey(id)), bs)
+		}
+	}()
+	return s.updateSecret(map[string][]byte{string(id): bs}, s.secretName)
 }
 
-// WriteTLSCertAndKey writes a TLS cert and key to domain.crt, domain.key fields of a Tailscale Kubernetes node's state
-// Secret.
-func (s *Store) WriteTLSCertAndKey(domain string, cert, key []byte) error {
-	return s.updateStateSecret(map[string][]byte{domain + ".crt": cert, domain + ".key": key})
+// WriteTLSCertAndKey writes a TLS cert and key to domain.crt, domain.key fields
+// of a Tailscale Kubernetes node's state Secret.
+func (s *Store) WriteTLSCertAndKey(domain string, cert, key []byte) (err error) {
+	defer func() {
+		if err == nil {
+			s.memory.WriteState(ipn.StateKey(sanitizeKey(domain+".crt")), cert)
+			s.memory.WriteState(ipn.StateKey(sanitizeKey(domain+".key")), key)
+		}
+	}()
+	secretName := s.secretName
+	data := map[string][]byte{domain + ".crt": cert, domain + ".key": key}
+	// If we run in cert share mode, cert and key for a DNS name are written
+	// to a separate Secret.
+	if envknob.IsCertShareReadWriteMode() {
```

Comment:
Should we return an error if we're in readonly mode?

Reply:
Theoretically, that would be the correct thing to do, but in practice if further down the line someone uses it in some sort of custom setup and misconfigures it so that a cert and key actually got issued in readonly mode, I'd rather store it and lessen the chance of them hitting the limits than error out because it was incorrect to issue in this mode and throw the cert away. I can add a log line